File: LC 1469. Find All The Lonely Nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> getLonelyNodes(TreeNode* root) {
        vector<int> ret;
        map<TreeNode*, vector<TreeNode*>> parent;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            TreeNode* tmp = q.front();
            q.pop();
            if(tmp->left){
                q.push(tmp->left);
                parent[tmp].push_back(tmp->left);
            }
            if(tmp->right){
                q.push(tmp->right);
                parent[tmp].push_back(tmp->right);
            }
        }
        for(auto i:parent){
            if(i.second.size() == 1){
                ret.push_back(i.second[0]->val);
            }
        }
        return ret;
    }
};

File: LC 1469. Find All The Lonely Nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> getLonelyNodes(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        HashMap<TreeNode, List<TreeNode>> parent = new HashMap<>();
        q.add(root);
        while(!q.isEmpty()){
            TreeNode tmp = q.poll();
            if(tmp.left != null){
                q.add(tmp.left);
                List<TreeNode> listVal = new ArrayList<>();
                if(parent.containsKey(tmp)){
                    listVal = parent.get(tmp);
                }
                listVal.add(tmp.left);
                parent.put(tmp, listVal);
            }
            if(tmp.right != null){
                q.add(tmp.right);
                List<TreeNode> listVal;
                if(parent.containsKey(tmp)){
                    listVal = parent.get(tmp);
                    listVal.add(tmp.right);
                }
                else{
                    listVal = new ArrayList<>();
                    listVal.add(tmp.right);
                }
                parent.put(tmp, listVal);
            }
        }
        for(TreeNode key:parent.keySet()){
            if(parent.get(key).size() == 1){
                list.add(parent.get(key).get(0).val);
            }
        }
        return list;
    }
}
